controllers/categories/services/update-bookmarks.service.ts

```diff
@@ -0,0 +1,136 @@
+import { Request, Response } from "express";
+import { Reference } from "@firebase/database-types";
+import { ResponseHandler } from "../../../services";
+
+// update type enum
+enum bookmarkUpdateType {
+  ADD = "add",
+  DELETE = "delete",
+}
+export async function updateCategoryBookmarks(
+  req: Request,
+  res: Response,
+  usersRef: Reference
+) {
+  try {
+    // user must have a session
+    const userId = req.session.userId,
+      categoryId = req.params.categoryId,
+      updateType: bookmarkUpdateType = req.body.updateType,
+      bookmarksToUpdate: [] = req.body.bookmarks,
+      // get a ref to the bookmarks object
+      bookmarksRef = usersRef.child(`${userId}/bookmarks`);
+
+    // Category must exist before any operations
+    await usersRef
+      .child(`${userId}/categories/${categoryId}`)
+      .once("value", async (categorySnapshot) => {
+        if (categorySnapshot.exists()) {
+          // push the bookmarks to the bookmarks object
+          await bookmarksRef.once(
+            "value",
+            async (bookmarksSnapshot) => {
+              if (updateType === bookmarkUpdateType.ADD) {
+                bookmarksToUpdate.forEach(async (bookmark) => {
+                  bookmarksRef.push(
+                    {
+                      categoryId,
+                      tweetId: bookmark,
+                    },
+                    (err) => {
+                      if (err) {
+                        // TODO: log to loggin service
+                        console.log(
+                          `Error creating bookmarks in category. see below: \n ${err}`
+                        );
+                        return ResponseHandler.clientError(
+                          res,
+                          "Error creating bookmarks in category"
+                        );
+                      }
+                    }
+                  );
+                });
+                // return success message
+                return ResponseHandler.requestSuccessful({
+                  res,
+                  message: "bookmarks added successfully",
+                  status: 201,
+                });
+                // check the update type
+              } else if (updateType === bookmarkUpdateType.DELETE) {
+                if (bookmarksSnapshot.exists()) {
+                  let ERROR_FLAG = false;
+                  // traverse through the bookmarks
+                  bookmarksSnapshot.forEach((bookmark) => {
+                    // check if the category id matches the provided category
+                    bookmarksToUpdate.forEach(async (bookmarkToDelete) => {
+                      if (
+                        bookmark.child("categoryId").val() === categoryId &&
```

Comment:
We can assign `bookmark.child("categoryId").value()` to a variable so we can shorten the if declaration on this line.

Reply:
The `"tweetId"` too.